zebra-state/src/service/memory_state.rs

```diff
@@ -460,11 +402,199 @@ impl Ord for Chain {
+                for chain in chains {
+                    self.chain_set.insert(chain);
+                }
+
+                // and return the chain
+                return Some(chain);
+            } else {
+                // add the chain back to the set and continue
+                self.chain_set.insert(chain);
+            }
+        }
+
+        None
+    }
+}
+
+/// A queue of blocks, awaiting the arrival of parent blocks.
+#[derive(Default)]
+pub struct QueuedBlocks {
+    /// Blocks awaiting their parent blocks for contextual verification.
+    blocks: HashMap<block::Hash, QueuedBlock>,
+    /// Hashes from `queued_blocks`, indexed by parent hash.
+    by_parent: HashMap<block::Hash, HashSet<block::Hash>>,
+    /// Hashes from `queued_blocks`, indexed by block height.
+    by_height: BTreeMap<block::Height, HashSet<block::Hash>>,
+}
+
+impl QueuedBlocks {
+    pub fn queue(&mut self, new: QueuedBlock) {
+        let new_hash = new.block.hash();
+        let new_height = new
+            .block
+            .coinbase_height()
+            .expect("validated non-finalized blocks have a coinbase height");
+        let parent_hash = new.block.header.previous_block_hash;
+
+        let replaced = self.blocks.insert(new_hash, new);
+        assert!(replaced.is_none(), "hashes must be unique");
+        let inserted = self
+            .by_height
+            .entry(new_height)
+            .or_default()
+            .insert(new_hash);
+        assert!(inserted, "hashes must be unique");
+        let inserted = self
+            .by_parent
+            .entry(parent_hash)
+            .or_default()
+            .insert(new_hash);
+        assert!(inserted, "hashes must be unique");
+
+        tracing::trace!(num_blocks = %self.blocks.len(), %parent_hash, ?new_height,  "Finished queueing a new block");
+    }
+
+    #[instrument(skip(self))]
+    pub fn dequeue_children(&mut self, parent: block::Hash) -> Vec<QueuedBlock> {
+        let queued_children = self
+            .by_parent
+            .remove(&parent)
+            .unwrap_or_default()
+            .into_iter()
+            .map(|hash| {
+                self.blocks
+                    .remove(&hash)
+                    .expect("block is present if its hash is in by_parent")
+            })
+            .collect::<Vec<_>>();
+
+        for queued in &queued_children {
+            let height = queued.block.coinbase_height().unwrap();
+            self.by_height.remove(&height);
+        }
+
+        tracing::trace!(num_blocks = %self.blocks.len(), "Finished dequeuing blocks waiting for parent hash",);
+
+        queued_children
+    }
+
+    pub fn prune_by_height(&mut self, finalized_tip_height: block::Height) {
+        // split_off returns the values _above_ the key, so we have to swap
+        // after to get the ones blow the key
+        let mut by_height = self.by_height.split_off(&finalized_tip_height);
+        mem::swap(&mut self.by_height, &mut by_height);
```

Comment:
The current code doesn't prune the finalized tip height. (But it should.)

Ideally, we could use`range` instead of `split_off` - but `range` returns an iterator, rather than splitting the set.

Is there a test for pruning the finalized tip height?
If there is, that test is also incorrect.
If not, please add one.

```suggestion
        // split_off returns the values _greater than or equal to_ the key, so we have to
        // add one to get the first non-finalized height, then swap to get the heights
        // less than or equal to the finalized tip height
        let mut by_height = self.by_height.split_off(&(finalized_tip_height + 1));
        mem::swap(&mut self.by_height, &mut by_height);
```